### PR TITLE
Add deletion secret as annotation to content

### LINKED ISSUE #161: Improve Deletion Secret Handling
This is the same underlying issue as https://github.com/kubernetes-csi/external-provisioner/issues/330 for external-provisioner. This bug is to track the work for the external-snapshotter. We should make sure the annotations used for both solution have the same name.

Today, the external-snapshotter will look at VolumeSnapshotClass for deletion secret. If the secret does not exist, delete is called any way. 

The problem is that the VolumeSnapshotClass may be deleted or mutated (deleted and recreated with different parameters). This will result in snapshots being unable to delete. Ideally we want the deletion secret on the VolumeSnapshotContent object. However, @liggitt pointed out that this would result in asymmetry of the API (provision is handled by a higher layer controller and specified in VolumeSnapshotClass, having that controller then look at VolumeSnapshotContent object for the delete secret seems wrong). That said, we do want to better handle this case.

So as a compromise, the proposal is to 1) add a reference to the deletion secret as an annotation on the VolumeSnapshotContent object (instead of a first class field), and 2) to better document why you shouldn't have deletion secrets.

For 1, the *proposed change* is to add a new flag to external-snapshotter that says controller requires deletion secret, if this is set by SP, the external-snapshotter should store a reference to the provision secret in an annotation on the PV object, and when deleting, if the flag is set and the VolumeSnapshotContent object has the annotation, fetch and pass the secret in the CSI `DeleteSnapshot` call.

And 2 is being tacked in https://github.com/kubernetes-csi/docs/issues/189#issuecomment-520621418.

CC @jingxu97 @xing-yang

### PR DESCRIPTION
**Pull request by xing-yang**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR adds deletion secret as annotation to volume snapshot content.
It also disallows annotation in template for snapshotter-secret-name.

**Which issue(s) this PR fixes**:
Fixes #161 #155 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Adds deletion secret as annotation to volume snapshot content.
```

### CODE REVIEW

File: pkg/controller/snapshot_controller.go [hunk 1]
```diff
@@ -687,6 +692,17 @@ func (ctrl *csiSnapshotController) createSnapshotOperation(snapshot *crdv1.Volum
 			DeletionPolicy:          class.DeletionPolicy,
 		},
 	}
+
+	// Set AnnDeletionSecretRefName and AnnDeletionSecretRefNamespace
+	if !metav1.HasAnnotation(snapshotContent.ObjectMeta, AnnDeletionSecretRefName) && snapshotterSecretRef != nil {
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/controller/snapshot_controller.go]: Sure.

File: pkg/controller/framework_test.go [hunk 2]
```diff
@@ -1287,6 +1288,27 @@ func secret() *v1.Secret {
 	}
 }
 
+func secretAnnotations() map[string]string {
+	return map[string]string{
+		AnnDeletionSecretRefName:      "secret",
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/controller/framework_test.go]: Yes, it is on line 1280.

File: pkg/controller/framework_test.go [hunk 3]
```diff
@@ -1287,6 +1288,27 @@ func secret() *v1.Secret {
...
+}
+
+func emptySecretAnnotations() map[string]string {
+	return map[string]string{
+		AnnDeletionSecretRefName:      "emptysecret",
+		AnnDeletionSecretRefNamespace: "default",
+	}
+}
+
+func invalidSecretAnnotations() map[string]string {
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/controller/framework_test.go]: Actually it is missing.

File: pkg/controller/framework_test.go [hunk 4]
```diff
@@ -1287,6 +1288,27 @@ func secret() *v1.Secret {
...
 }
 
+func secretAnnotations() map[string]string {
+	return map[string]string{
+		AnnDeletionSecretRefName:      "secret",
+		AnnDeletionSecretRefNamespace: "default",
+	}
+}
+
+func emptySecretAnnotations() map[string]string {
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/controller/framework_test.go]: Ok

File: pkg/controller/snapshot_controller.go [hunk 5]
```diff
@@ -736,23 +752,29 @@ func (ctrl *csiSnapshotController) deleteSnapshotContentOperation(content *crdv1
...
+	var err error
+
+	// Check if annotation exists
+	if metav1.HasAnnotation(content.ObjectMeta, AnnDeletionSecretRefName) && metav1.HasAnnotation(content.ObjectMeta, AnnDeletionSecretRefNamespace) {
+		annDeletionSecretName := content.Annotations[AnnDeletionSecretRefName]
+		annDeletionSecretNamespace := content.Annotations[AnnDeletionSecretRefNamespace]
+
+		snapshotterSecretRef := &v1.SecretReference{}
+
+		if annDeletionSecretName == "" || annDeletionSecretNamespace == "" {
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/controller/snapshot_controller.go]: Ok

File: pkg/controller/framework_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ggriffiths) [reply pkg/controller/framework_test.go]: In this case, `emptysecret` exists, but it does not have any data inside. I'll update the name to `emptyDataSecretAnnotations` to be more clear.

(I'm helping Xing with these tests)

File: pkg/controller/snapshot_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ggriffiths) [reply pkg/controller/snapshot_controller.go]: Updated unit test 1-4 to use `emptyNamespaceSecretAnnotations()`

File: pkg/controller/framework_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ggriffiths) [reply pkg/controller/framework_test.go]: removed this.

File: pkg/controller/snapshot_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ggriffiths) [reply pkg/controller/snapshot_controller.go]: Removed this check. Still checking `if snapshotterSecretRef != nil` as we ran into a nil pointer without it.

**Comment by ggriffiths** [comment]
Addressed feedback and squashed to one commit. @msau42

File: pkg/controller/snapshot_delete_test.go [hunk 6]
```diff
@@ -165,8 +165,8 @@ func TestDeleteSync(t *testing.T) {
 		},
 		{
 			name:              "1-4 - fail delete with snapshot class that has invalid secret parameter",
-			initialContents:   newContentArray("content1-4", invalidSecretClass, "sid1-4", "vuid1-4", "volume1-4", "snapuid1-4", "snap1-4", &deletePolicy, nil, nil, true),
-			expectedContents:  newContentArray("content1-4", invalidSecretClass, "sid1-4", "vuid1-4", "volume1-4", "snapuid1-4", "snap1-4", &deletePolicy, nil, nil, true),
+			initialContents:   newContentArray("content1-4", invalidSecretClass, "sid1-4", "vuid1-4", "volume1-4", "snapuid1-4", "snap1-4", &deletePolicy, nil, nil, true, emptyNamespaceSecretAnnotations()),
+			expectedContents:  newContentArray("content1-4", invalidSecretClass, "sid1-4", "vuid1-4", "volume1-4", "snapuid1-4", "snap1-4", &deletePolicy, nil, nil, true, emptyNamespaceSecretAnnotations()),
```
Comment (msau42) [inline pkg/controller/snapshot_delete_test.go]: Is there a test case for if a secret is specified, but it doesn't exist?  To mirror the external-provisioner scenario, I believe we continue and pass no secret to the Delete call, and it's up to the plugin to fail if they need it.

Comment (ggriffiths) [reply pkg/controller/snapshot_delete_test.go]: Unless I'm missing something, it seems that the current behavior for the snapshotter is to fail in this case:
https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/controller/snapshot_controller.go#L748-L751

Comment (xing-yang) [reply pkg/controller/snapshot_delete_test.go]: @ggriffiths I missed the provisioner behavior in this case.  Please go ahead to fix it and add a test case.  Thanks.

Comment (ggriffiths) [reply pkg/controller/snapshot_delete_test.go]: Fixed and added deletion test `1-16` for continuing deletion. Also added a test for creation: `7-10`

**Comment by msau42** [comment]
/lgtm
Thanks!

**Pull request merged**